narwhal/consensus/src/consensus.rs

```diff
@@ -31,6 +35,114 @@ pub mod consensus_tests;
 /// The representation of the DAG in memory.
 pub type Dag = BTreeMap<Round, HashMap<AuthorityIdentifier, (CertificateDigest, Certificate)>>;
 
+#[derive(Default)]
+pub struct LeaderSwapTable {
+    pub good_nodes: Vec<AuthorityIdentifier>,
+    pub bad_nodes: HashSet<AuthorityIdentifier>,
+}
+
+impl LeaderSwapTable {
+    // constructs a new table based on the provided reputation scores.
+    pub fn new(committee: &Committee, reputation_scores: ReputationScores) -> Self {
+        assert!(reputation_scores.final_of_schedule, "Only reputation scores that have been calculated on the end of a schedule are accepted");
+
+        let good_nodes = Self::good_nodes(&reputation_scores, committee);
+        let bad_nodes = Self::bad_nodes(&reputation_scores, committee);
+
+        debug!("Reputation scores: {:?}", reputation_scores);
+        debug!("Good nodes: {:?}", good_nodes);
+        debug!("Bad nodes: {:?}", bad_nodes);
+
+        Self {
+            good_nodes,
+            bad_nodes,
+        }
+    }
+
+    /// Checks whether the provided leader is a bad performer and needs to be swapped for the schedule
+    /// with a good performer. If not, then the method returns None. Otherwise the leader to swap with
+    /// is returned instead. The `leader_round` represents the DAG round on which the provided AuthorityIdentifier
+    /// is a leader on and is used as a seed to random function in order to calculate the good node that
+    /// will swap for that round the bad node. We are intentionally not doing weighted randomness as
```

Comment:
> We are intentionally not doing weighted randomness as we want to give to all the good nodes equal opportunity to get swapped with bad nodes and not have one node with enough stake dominating on the final schedule.

I can see using weighted random among good nodes can skew good nodes with larger stakes to have disproportionally more voting power. But I don't see how this can make one node dominate. More details?

Reply:
Probably poor choice of word "dominating" from my side. Basically we don't want to have a node with high stake occupying a large portion of the schedule slots of the `f` bad nodes.